### ISSUE TITLE
Black card doesn't always appear; chaos ensues

### ISSUE THREAD

**Issue by pao** [berrycardsagainsthumanity/cah#2]
Well, this is why we're all here, isn't it?

**Comment by pao** [comment]
``` Python
        return {"text": black_card.text,
                "card_id": black_card.id,
                "num_white_cards": 1 if not black_card.pick else black_card.pick}
```

https://github.com/berrycardsagainsthumanity/cah/blob/master/webroot/game.py#L263

Are there queries for which one of the values in here might be `None`?

**Comment by pao** [comment]
```
<Marminatoror>  First thing I do is grab all the black card and white card ids
    And unless the database was changing during
    which it's not
    I think that the query by id should always work
    Also
    I think that num_white_cards is availalb
    Or rather
    num_white_cards is available
    so the issue must be later than that
    black_card = self._get_black_card()
    self._state.black_card = black_card
    self._publish("add_black", black_card)
<pao>   and all rows in the table must have id NOT NULL by construction
```

**Comment by pao** [comment]
```
<pao>   does the <div> make it into the layout when the card is missing? (not that you've had the chance to look)
<Marminatoror>  Dunni
    dunno
<pao>   but if it's missing completely, then that would imply the JS that adds it never runs
<Marminatoror>  but I would guess that the div doesn't make it there
    because if the div were there it would display
    even if the text was blank
<pao>   right, there should be a big black nothing
```

**Comment by pao** [comment]
```
<Marminatoror>  Oh, could also be some kind of race condition
    Although I don't see how
    Well hmmmmmm
<pao>   it's async, so it could always be a race, but what resource would race?
<Marminatoror>  I guess I don't know much about how the threading stuff works in python with twisted
<pao>   no thread, just coroutines
<Marminatoror>  I figured
<pao>   let me check that though
<Marminatoror>  _start_round() called twice at the same time or something?
    Hmm
    actually I have no idea how that would happen
    _start_round should only be called by the 10 second timer
    If old czar left the game at the same time as the timer firing?
```

**Comment by berrycardsagainsthumanity** [comment]
Here is a log file from a black card that didn't show up. No exceptions. I took a look at the javascript console and the same time and the add_black message doesn't appear. This means the card is properly being retrieved AND published, but doesn't make it to the users for some reason.

2013-01-19 18:17:51,392 INFO sqlalchemy.engine.base.Engine /home/ubuntu/.virtualenvs/cah/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:851 SELECT cards.id AS cards_id, cards.text AS cards_text, cards.cardset AS cards_cardset, cards.draw AS cards_draw, cards.pick AS cards_pick, cards.is_black_card AS cards_is_black_card
FROM cards
WHERE cards.id = %(id_1)s
 LIMIT %(param_1)s
2013-01-19 18:17:51,392 INFO sqlalchemy.engine.base.Engine /home/ubuntu/.virtualenvs/cah/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py:853 {'id_1': 966, 'param_1': 1}
2013-01-19 18:17:51,394 INFO cah.game /NotWork/cah/webroot/game.py:314 Publishing: add_black, Data: {'text': u'Giggle at ____!', 'card_id': 966, 'num_white_cards': 1}, exclude: None, eligible: None
2013-01-19 18:17:51,394 INFO cah.game /NotWork/cah/webroot/game.py:314 Publishing: czar_chosen, Data: someone, exclude: None, eligible: None

**Comment by pao** [comment]
And it's not the specific card, because I saw that one pop up during gameplay yesterday. I do see that sqlalchemy outputs Unicode strings from the database.

**Comment by pao** [comment]
Possibly fixed by 5dc0fb9d77beba4dd7681f2164716827fd6a3505?

**Comment by pao** [comment]
Let's assume this is fixed based on playtesting last night.
